[FIX] hr_holidays: Prevent Time Off Without Allocation (#176149)

MohdAlmosawy opened the pull request:
This commit addresses an issue where time off requests could be validated even when the requested dates fall outside the allocated leave period. The `_check_validity` method now correctly checks the allocation dates and raises a validation error when the dates are outside the allocation period, ensuring the integrity of time off requests.

Fixes #176149

**Description of the issue/feature this PR addresses:**
The issue addressed by this PR is that time off requests in draft state could be validated even when the requested dates fall outside the allocated leave period. This behavior leads to inconsistencies in leave management, allowing invalid requests to bypass the allocation check.

**Current behavior before PR:**
Time off requests in draft state could be validated without checking if the requested dates are within the allocated leave period, which can result in employees being granted leave that they are not entitled to.

**Desired behavior after PR is merged:**
The _check_validity method will correctly check the allocation dates for time off requests in draft state. If the requested dates fall outside the allocated leave period, a validation error will be raised, ensuring that only valid time off requests are approved.


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/176430.png)](https://mergebot.odoo.com/odoo/odoo/pull/176430)

ChristophAbenthungCibex commented on addons/hr_holidays/models/hr_leave.py:
```diff
@@ -784,6 +784,17 @@ def _check_validity(self):
+                ('employee_id', 'in', employees.ids),
+                ('state', '=', 'validate')
+            ])
+            # Validate against the allocation period
+            for allocation in allocation_records:
+                if leave.request_date_from < allocation.date_from or leave.request_date_to > allocation.date_to:
+                    raise ValidationError(_("The leave dates are outside the allocation period."))
```
When the employee has an allocation from January 1st with end date at January 31st (only valid in January) and then a second allocation starting from February 2nd and no end date. Both are validated. The employee requests a time off for March 1st. That leave request is valid because the employee has an allocation that is valid for March.
As far as I can tell this code would throw a ValidationError because there is an allocation outside of the requested leave. Wouldn't it be better to add the start and end date to the domain and check if any allocation is loaded from the database?

MohdAlmosawy commented:
Hi @ChristophAbenthungCibex,

Thank you for your helpful feedback! I have updated the _check_validity method to cover your concerns:

- Relevant Allocations: I changed the allocation search to only look for relevant allocations, and now it also works with open-ended dates (date_from or date_to being False).
- Handling Overlaps: The new logic checks for any valid allocation for the leave request, so it won't throw an error if other allocations are outside the requested period.
Better Validation: I improved the logic to ensure it handles all cases correctly, including those with missing allocation dates.

ChristophAbenthungCibex commented on addons/hr_holidays/models/hr_leave.py:
```diff
@@ -784,21 +784,47 @@ def _check_validity(self):
+                ('date_from', '=', False),
+                ('date_from', '<=', max(leave.request_date_from for leave in leaves))
+            ])
+            # Check for any valid allocation covering the leave request period
+            valid_allocation = any(
+                allocation.date_from <= leave.request_date_from <= leave.request_date_to <= allocation.date_to
+                for leave in leaves
+                for allocation in allocation_records
+                if allocation.date_from and allocation.date_to
+            )
```
Hello @MohdAlmosawy 
Thanks for the improvements. My previous concerns are now fixed.
A minor Issue that I found:
An employee has a valid allocation that has date_from set as 1st August and no end_date (allocation does not expire).
The employee requests a time off during August. The ValidationError is thrown because the only valid leave is filtered out due to missing a date_to.
Maybe change it to something like the code below. Add "defaults" to the date comparison then there is no need to only check the allocations that have date_from and date_to set. 
```suggestion
            # Check for any valid allocation covering the leave request period
            valid_allocation = any(
                allocation.date_from or datetime.min <= leave.request_date_from <= leave.request_date_to <= allocation.date_to or datetime.max
                for leave in leaves
                for allocation in allocation_records
            )
```

ChristophAbenthungCibex reviewed (commented):
@MohdAlmosawy 
LGTM
Thank you

Bertrand2 commented:
Hello, thanks for your PR!

Concerning the bug mentioned in the PR message about the draft state, we already have an ongoing fix in PR odoo/odoo#176936 addressing it.
 Concerning @ChristophAbenthungCibex 's comments, I'm not sure if it is related to the PR changes or if the issues are related to the current code of odoo. As far as I can tell it's related to the PR, but I prefer to ask just to be sure :)
 
 Have a nice day!

ChristophAbenthungCibex commented:
Hi @Bertrand2 
My concerns mentioned in this Pull Request are as you expected for the changes of this Pull Request. 
Regarding the Pull Request you have mentioned (#176936): I do not see how this Pull Request will fix the Issue #176149. As far as I can tell is the validation itself broken and not that it is not triggered / executed properly. Maybe I am wrong or missing something. Please let me know, thank you.

Bertrand2 reviewed (commented):
@ChristophAbenthungCibex indeed, I misunderstood the commit/PR message, my bad. :slightly_smiling_face: 

Without diving in the details of the code itself, here are already a first review:
- It would be nice to squash the commits into one to make it clearer
- If possible, include the steps mentioned in the issue back in the commit/PR message as this is something that we usually include when making a fix.
- Concerning the code, the way the check is done isn't right. We might have a situation for example at the end of the year, where one allocation ends December 31st while the other starts the day after, January 1st. With this check, the issue would be that it wouldn't be possible to take a leave on both days, while it should be possible.
- Finally, the simplest solutions that I visualize for that particular issue might mess up other stuff; I think of our french localization for time off that needs to adapt date_from and date_to, which can make the dates fall outside out allocation validity while still being valid :thinking: 

Do you mind if I give that bug fix to a developer from my team so that we can think a bit more on all the implications? I believe the fix isn't that urgent as this is probably a very specific edge case.

Thanks, have a great day! :smile:
